pkg/networkserver/downlink.go

```diff
@@ -803,6 +809,227 @@ const gsScheduleWindow = 30 * time.Second
+
+	case ttnpb.MType_REJOIN_REQUEST:
+		rxDelay = ttnpb.RxDelay(phy.JoinAcceptDelay1 / time.Second)
+
+	default:
+		logger.Warn("Last uplink is neither data uplink, nor rejoin-request, skip class A downlink slot")
+		dev.MACState.QueuedResponses = nil
+		dev.MACState.RxWindowsAvailable = false
+		return downlinkAttemptResult{
+			SetPaths: append(sets,
+				"mac_state.queued_responses",
+				"mac_state.rx_windows_available",
+			),
+		}
+	}
+	ctx = events.ContextWithCorrelationID(ctx, up.CorrelationIDs...)
+
+	rx1, rx2, paths := downlinkPathsForClassA(rxDelay, dev.RecentUplinks...)
+	if !rx1 && !rx2 {
+		logger.Warn("Rx1 and Rx2 are expired, skip class A downlink slot")
+		dev.MACState.QueuedResponses = nil
+		dev.MACState.RxWindowsAvailable = false
+		return downlinkAttemptResult{
+			SetPaths: append(sets,
+				"mac_state.queued_responses",
+				"mac_state.rx_windows_available",
+			),
+		}
+	}
+	if len(paths) == 0 {
+		logger.Warn("No downlink path available, skip class A downlink slot")
+		return downlinkAttemptResult{
+			SetPaths: sets,
+		}
+	}
+	if dev.MACState.DeviceClass != ttnpb.CLASS_A && !rx1 && rx2 && len(dev.MACState.QueuedResponses) == 0 {
+		logger.Debug("Skip Rx2-only Class A downlink for non-class A device")
+		return downlinkAttemptResult{
+			SetPaths: sets,
+		}
+	}
+
+	req, err := txRequestFromUplink(phy, dev.MACState, rx1, rx2, rxDelay, up)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to generate Tx request from uplink, skip class A downlink slot")
+		return downlinkAttemptResult{
+			SetPaths: sets,
+		}
+	}
+
+	var maxDR ttnpb.DataRateIndex
+	switch {
+	case rx1 && rx2:
+		maxDR = req.Rx1DataRateIndex
+		if req.Rx2DataRateIndex > maxDR {
+			maxDR = req.Rx2DataRateIndex
+		}
+
+	case rx1:
+		maxDR = req.Rx1DataRateIndex
+
+	case rx2:
+		maxDR = req.Rx2DataRateIndex
+	}
+
+	genDown, genState, err := ns.generateDownlink(ctx, dev, phy, ttnpb.CLASS_A,
+		phy.DataRates[maxDR].DefaultMaxSize.PayloadSize(fp.DwellTime.GetDownlinks()),
+		maxUpLength,
+	)
+	if genState.NeedsDownlinkQueueUpdate {
+		sets = append(sets,
+			"queued_application_downlinks",
+		)
+	}
+	if err != nil {
+		switch {
+		case errors.Resemble(err, errNoDownlink):
+			logger.Debug("No class A downlink to send, skip class A downlink slot")
+
+		default:
+			logger.WithError(err).Warn("Failed to generate class A downlink, skip class A downlink slot")
+		}
+		if genState.ApplicationDownlink != nil {
+			dev.QueuedApplicationDownlinks = append([]*ttnpb.ApplicationDownlink{genState.ApplicationDownlink}, dev.QueuedApplicationDownlinks...)
+		}
+		return downlinkAttemptResult{
+			SetPaths:              sets,
+			applicationUpAppender: genState.appendApplicationUplinks,
+		}
+	}
+
+	if rx1 && rx2 {
+		rx1 = len(genDown.Payload) <= int(phy.DataRates[req.Rx1DataRateIndex].DefaultMaxSize.PayloadSize(fp.DwellTime.GetDownlinks()))
+		rx2 = len(genDown.Payload) <= int(phy.DataRates[req.Rx2DataRateIndex].DefaultMaxSize.PayloadSize(fp.DwellTime.GetDownlinks()))
```

Comment:
Changing the meaning of the variables here can cause confusion in this block

Reply:
The meaning of the variables does not change - in both cases it's `is rx{1,2} possible?`

Reply:
OK